File: src/camera/ActionDetector.ts
import { FaceDetectionResult } from "./FaceDetector";

export interface ActionState {
  action: string;
  detected: boolean;
  active: boolean;
  confidence: number;
}

/**
 * Per-action detection tuning.
 * - threshold : minimum confidence score required to start counting.
 * - holdFrames: consecutive frames above threshold before action is confirmed.
 *
 * BLINK is a fast reflex (~100-200ms) so we use a low threshold and only 2
 * frames. Head-turn and mouth-open are sustained movements so they need a
 * slightly longer hold to avoid false positives.
 */
const ACTION_CONFIGS: Record<string, { threshold: number; holdFrames: number }> = {
  BLINK:        { threshold: 0.30, holdFrames: 2 },
  TURN_LEFT:    { threshold: 0.45, holdFrames: 6 },
  TURN_RIGHT:   { threshold: 0.45, holdFrames: 6 },
  TURN_HEAD:    { threshold: 0.45, holdFrames: 6 },
  OPEN_MOUTH:   { threshold: 0.40, holdFrames: 3 },
};
const DEFAULT_CONFIG = { threshold: 0.55, holdFrames: 8 };

/**
 * Sequential action detector aligned with backend's 5 UPPERCASE actions:
 *   BLINK, TURN_LEFT, TURN_RIGHT, TURN_HEAD, OPEN_MOUTH
 */
export class ActionDetector {
  private completedActions = new Set<string>();
  private holdCounters = new Map<string, number>();

  constructor(private requiredActions: string[]) {}

  check(result: FaceDetectionResult): ActionState[] {
    const currentAction = this.getCurrentAction();

    if (!result.faceDetected) {
      return this.buildStates(currentAction, 0);
    }

    const bs = result.blendshapes;

    // Only check the current active action (sequential flow)
    if (currentAction && !this.completedActions.has(currentAction)) {
      const conf = this.getConfidence(currentAction, bs, result.transformMatrix);
      const cfg = ACTION_CONFIGS[currentAction] ?? DEFAULT_CONFIG;

      if (conf >= cfg.threshold) {
        const count = (this.holdCounters.get(currentAction) || 0) + 1;
        this.holdCounters.set(currentAction, count);

        if (count >= cfg.holdFrames) {
          this.completedActions.add(currentAction);
          this.holdCounters.delete(currentAction);
        }
      } else {
        this.holdCounters.set(currentAction, 0);
      }

      return this.buildStates(currentAction, conf);
    }

    return this.buildStates(currentAction, 0);
  }

  getCurrentAction(): string | null {
    for (const action of this.requiredActions) {
      if (!this.completedActions.has(action)) return action;
    }
    return null;
  }

  private buildStates(
    currentAction: string | null,
    currentConf: number
  ): ActionState[] {
    return this.requiredActions.map((action) => ({
      action,
      detected: this.completedActions.has(action),
      active: action === currentAction,
      confidence: action === currentAction ? currentConf : 0,
    }));
  }

  private getConfidence(
    action: string,
    bs: Map<string, number>,
    matrix: number[] | null
  ): number {
    switch (action) {
      case "BLINK": {
        const left = bs.get("eyeBlinkLeft") || 0;
        const right = bs.get("eyeBlinkRight") || 0;
        return (left + right) / 2;
      }

      case "TURN_LEFT": {
        const yaw = this.getHeadYaw(matrix);
        return yaw < -12 ? Math.min(1, Math.abs(yaw) / 30) : 0;
      }

      case "TURN_RIGHT": {
        const yaw = this.getHeadYaw(matrix);
        return yaw > 12 ? Math.min(1, yaw / 30) : 0;
      }

      case "TURN_HEAD": {
        // Any significant yaw (left OR right) counts
        const yaw = Math.abs(this.getHeadYaw(matrix));
        return yaw > 15 ? Math.min(1, yaw / 30) : 0;
      }

      case "OPEN_MOUTH":
        return bs.get("jawOpen") || 0;

      default:
        return 0;
    }
  }

  private getHeadYaw(matrix: number[] | null): number {
    if (!matrix || matrix.length < 16) return 0;
    return Math.atan2(matrix[8], matrix[0]) * (180 / Math.PI);
  }

  allCompleted(): boolean {
    return this.requiredActions.every((a) => this.completedActions.has(a));
  }

  completedCount(): number {
    return this.completedActions.size;
  }

  reset(): void {
    this.completedActions.clear();
    this.holdCounters.clear();
  }
}

File: flutter/lib/src/camera/action_detector.dart
import 'package:google_mlkit_face_detection/google_mlkit_face_detection.dart';
import '../types/liveness.dart';

/// Tracks the state of one liveness action.
class ActionState {
  final ChallengeAction action;
  bool detected;
  bool active;
  double confidence;

  ActionState({
    required this.action,
    this.detected = false,
    this.active = false,
    this.confidence = 0,
  });
}

/// Detects and sequences face actions using ML Kit face data.
/// Requires [_holdFrames] consecutive high-confidence frames before
/// marking an action as completed.
class ActionDetector {
  static const int _holdFrames = 8;
  static const double _threshold = 0.6;

  final List<ChallengeAction> requiredActions;
  final Set<ChallengeAction> _completed = {};
  final Map<ChallengeAction, int> _holdCounters = {};

  ActionDetector(this.requiredActions);

  ChallengeAction? getCurrentAction() {
    for (final action in requiredActions) {
      if (!_completed.contains(action)) return action;
    }
    return null;
  }

  List<ActionState> check(Face? face) {
    final current = getCurrentAction();

    if (face == null || current == null) {
      return _buildStates(current, 0);
    }

    if (!_completed.contains(current)) {
      final conf = _getConfidence(current, face);

      if (conf > _threshold) {
        final count = (_holdCounters[current] ?? 0) + 1;
        _holdCounters[current] = count;
        if (count >= _holdFrames) {
          _completed.add(current);
          _holdCounters.remove(current);
        }
      } else {
        _holdCounters[current] = 0;
      }

      return _buildStates(current, conf);
    }

    return _buildStates(current, 0);
  }

  bool allCompleted() =>
      requiredActions.every((a) => _completed.contains(a));

  void reset() {
    _completed.clear();
    _holdCounters.clear();
  }

  List<ActionState> _buildStates(ChallengeAction? current, double conf) {
    return requiredActions.map((a) => ActionState(
          action: a,
          detected: _completed.contains(a),
          active: a == current,
          confidence: a == current ? conf : 0,
        )).toList();
  }

  double _getConfidence(ChallengeAction action, Face face) {
    switch (action) {
      case ChallengeAction.blink:
        final l = 1.0 - (face.leftEyeOpenProbability ?? 1.0);
        final r = 1.0 - (face.rightEyeOpenProbability ?? 1.0);
        return (l + r) / 2;

      case ChallengeAction.turnLeft:
        final yaw = face.headEulerAngleY ?? 0.0;
        return yaw < -12 ? (yaw.abs() / 30).clamp(0.0, 1.0) : 0.0;

      case ChallengeAction.turnRight:
        final yaw = face.headEulerAngleY ?? 0.0;
        return yaw > 12 ? (yaw / 30).clamp(0.0, 1.0) : 0.0;

      case ChallengeAction.turnHead:
        // Accept either a left or right turn
        final yaw = (face.headEulerAngleY ?? 0.0).abs();
        return yaw > 12 ? (yaw / 30).clamp(0.0, 1.0) : 0.0;

      case ChallengeAction.openMouth:
        // ML Kit doesn't expose jawOpen — use a rough heuristic via smile
        // (confidence will be low; backend is the true arbiter via video)
        return face.smilingProbability ?? 0.0;
    }
  }
}
